src/searchModal/advancedSearch.js

```diff
@@ -0,0 +1,456 @@
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA ;
+ */
+import $ from 'jquery';
+import _ from 'lodash';
+import __ from 'i18n';
+import advancedSearchTpl from 'ui/searchModal/tpl/advanced-search';
+import textCriterionTpl from 'ui/searchModal/tpl/text-criterion';
+import invalidCriteriaWarningTpl from 'ui/searchModal/tpl/invalid-criteria-warning';
+import listCheckboxCriterionTpl from 'ui/searchModal/tpl/list-checkbox-criterion';
+import listSelectCriterionTpl from 'ui/searchModal/tpl/list-select-criterion';
+import 'ui/searchModal/css/advancedSearch.css';
+import component from 'ui/component';
+import 'ui/modal';
+import 'ui/datatable';
+import 'select2';
+import urlUtil from 'util/url';
+import request from 'core/dataProvider/request';
+
+/**
+ * Creates advanced search component
+ *
+ * @param {object} config
+ * @param {object} config.renderTo - DOM element where component will be rendered to
+ * @param {string} config.advancedCriteria - advanced criteria to be set on component creation
+ * @returns {advancedSearch}
+ */
+export default function advancedSearchFactory(config) {
+    // Private properties to be easily accessible by instance methods
+    let $container = null;
+    let $addCriteriaInput = null;
+    let $criteriaSelect = null;
+    let $advancedCriteriaContainer = null;
+    let criteriaState = null;
+
+    const criteriaTypes = {
+        text: 'text',
+        list: 'list'
+    };
+
+    // Creates new component
+    const instance = component({
+        /**
+         * Request metadata (criteria) for the given uri
+         * @param {string} classUri - Uri of the class to get metadata from
+         * @param {number} maxListSize - max number of values to return for a metadata of type list If
+         *                               available values options exceed this number, an uri will
+         *                               be returned instead to get its values dinamically
+         * @returns {Promise} - Request promise
+         */
+        updateCriteria: function (classUri, maxListSize = 5) {
+            const route = urlUtil.route('get', 'ClassMetadata', 'tao', { classUri, maxListSize });
```

Comment:
**thought:** Since this component is in a shared repo, the URL should not be hardcoded.
**suggestion:** Make it configurable.